keps/sig-node/3085-pod-conditions-for-starting-completition-of-sandbox-creation/README.md

```diff
@@ -0,0 +1,1392 @@
+where individual tenants own the pod specs (including the set of init
+containers) while the cluster administrators are in charge of storage plugins,
+networking plugins and container runtime handlers.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- Surface the beginning of pod sandbox creation by Kubelet as a pod condition.
+- Surface the completion of pod sandbox creation by Kubelet as a pod condition.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Modify the meaning of the existing Initialized condition
+- Specify metrics collection based on the conditions around pod sandbox creation
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+This KEP proposes enhancements to the Kubelet to report the start and completion
+of pod sandbox creation as two pod conditions with types:
+`PodSandboxCreationStarted` and `PodSandboxCreationCompleted`. Metric collection
+and monitoring services can use the fields associated with the two pod
+conditions to aggregate and report sandbox creation latencies (as well as
+failures to create the pod sandbox) based on various properties of the pod:
+number of volumes, storage class of PVCs, runtime class, custom annotations for
+CNI and IPAM plugins, etc.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+In each of the User Stories below, nearly identical pod conditions that would
+result from different scenarios/problems are grouped together. The unique
+scenarios are detailed after describing the pod conditions. To make the stories
+concrete, a specific set of timestamps in the future is chosen. The
+`PodScheduled` condition is mentioned in the stories but conditions after pod
+sandbox creation (e.g. `Initialized` and `Ready`) are skipped.
+
+#### Story 1: Stateless pod scheduled on a healthy node and cluster
+
+A user launches a simple, stateless runc based pod in a healthy cluster. The pod
+gets successfully scheduled at 2022-12-06T15:33:46Z, pod sandbox creation starts
+at 2022-12-06T15:33:47Z and completes successfully after two seconds at
+2022-12-06T15:33:49Z.
+
+The pod will report the following conditions in pod status at
+2022-12-06T15:33:48Z (after pod sandbox creation started but did not complete):
```

Comment:
What about initialization?

Reply:
Are you referring to what the `Initialized` pod condition would state below? If so, the `Initialized` pod condition would depend on whether the pod has `init` containers configured or not. I can add a note clarifying this here with sample values (but this fact is also called out [earlier](https://github.com/kubernetes/enhancements/pull/3087/files#diff-08c9a02aebe0409f9d02d84d2a337e642394319991f684d82798f400510a6d85R237))

Reply:
Will also add how conditions will look like before pod sandbox creation has started but after a kubelet work has started processing.

Reply:
Addressed [here](https://github.com/kubernetes/enhancements/pull/3087/files#diff-08c9a02aebe0409f9d02d84d2a337e642394319991f684d82798f400510a6d85R387) and [here](https://github.com/kubernetes/enhancements/pull/3087/files#diff-08c9a02aebe0409f9d02d84d2a337e642394319991f684d82798f400510a6d85R462)